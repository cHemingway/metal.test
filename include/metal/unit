#if defined(__cplusplus)
#include <metal/unit.hpp>
#else
#include <metal/unit.h>
#endif