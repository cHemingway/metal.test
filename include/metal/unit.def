/**
 * @file   metal/unit.def
 * @date   25.10.2016
 * @author Klemens D. Morgenstern
 *
 */

#ifndef METAL_UNIT_DEF_
#define METAL_UNIT_DEF_

typedef enum __metal_level_t
{
    __metal_level_assert,
    __metal_level_expect,
} __metal_level;

typedef enum __metal_oper_t
{
    __metal_oper_enter_case,
    __metal_oper_exit_case,
    __metal_oper_enter_ranged,
    __metal_oper_exit_ranged,
    __metal_oper_log,
    __metal_oper_checkpoint,
    __metal_oper_message,
    __metal_oper_plain,
    __metal_oper_predicate,
    __metal_oper_equal,
    __metal_oper_not_equal,
    __metal_oper_close,
    __metal_oper_close_rel,
    __metal_oper_close_perc,
    __metal_oper_ge,
    __metal_oper_greater,
    __metal_oper_le,
    __metal_oper_lesser,
    __metal_oper_exception,
    __metal_oper_any_exception,
    __metal_oper_no_exception,
    __metal_oper_no_exec,
    __metal_oper_exec,
    __metal_oper_report
} __metal_oper;

extern int __metal_status  ;
extern int __metal_critical;
extern int __metal_errored;

void __metal_call(void (*func)(), const char * msg,
               const char * file, int line);

void __metal_unit_break(__metal_level lvl,
               __metal_oper oper,
               int condition,
               int bitwise,
               const char* str1,
               const char* str2,
               const char* str3,
               const char* file,
               int line);

#if defined(METAL_UNIT_ONLY_ERRORS)
#define __metal_unit_impl(lvl, oper, condition, ...) (((condition == 0) && (lvl == __metal_level_assert)) ? __metal_unit_break(lvl, oper, condition, __VA_ARGS__) : ((void)0))
#elif defined(METAL_UNIT_ONLY_WARNINGS)
#define __metal_unit_impl(lvl, oper, condition, ...) ((condition == 0) ? __metal_unit_break(lvl, oper, condition, __VA_ARGS__) : ((void)0))
#else
#define __metal_unit_impl(...) __metal_unit_break(__VA_ARGS__)
#endif

#if defined(METAL_UNIT_ONLY_ERRORS) || defined(METAL_UNIT_ONLY_WARNINGS)
#define __metal_unit_call(Function, ...) Function()
#else
#define __metal_unit_call(...) __metal_call(__VA_ARGS__)
#endif

int __metal_report();

#define __METAL_BITWISE_STEP(Lhs, Rhs, Index, Oper) (((Lhs >> Index) & 1) Oper ((Rhs >> Index) & 1))

#define __METAL_BITWISE_8(Lhs, Rhs, Oper, Chain) \
    ((Lhs & 1) Oper (Rhs & 1))                   \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x4, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x5, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x6, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x7, Oper)

#define __METAL_BITWISE_16(Lhs, Rhs, Oper, Chain) \
    __METAL_BITWISE_8(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x8, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x9, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xA, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xB, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xC, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xD, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xE, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0xF, Oper)

#define __METAL_BITWISE_32(Lhs, Rhs, Oper, Chain) \
    __METAL_BITWISE_16(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x10, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x11, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x12, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x13, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x14, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x15, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x16, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x17, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x18, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x19, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x1F, Oper)

#define __METAL_BITWISE_64(Lhs, Rhs, Oper, Chain)    \
    __METAL_BITWISE_32(Lhs, Rhs, Oper, Chain)        \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x20, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x21, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x22, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x23, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x24, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x25, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x26, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x27, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x28, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x29, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x2F, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x30, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x31, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x32, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x33, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x34, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x35, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x36, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x37, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x38, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x39, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3A, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3B, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3C, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3D, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3E, Oper) \
    Chain __METAL_BITWISE_STEP(Lhs, Rhs, 0x3F, Oper)

#define __METAL_BITWISE_SIZE(Lhs, Rhs) ((sizeof(Lhs) >= sizeof(Rhs)) ? sizeof(Lhs) : sizeof(Rhs))

#define METAL_CONCAT_IMPL( x, y ) x##y
#define METAL_CONCAT( x, y ) METAL_CONCAT_IMPL( x, y )

#define METAL_STRINGIZE(x) METAL_STRINGIZE2(x)
#define METAL_STRINGIZE2(x) #x
#define METAL_LOCATION_STR() __FILE__ "(" METAL_STRINGIZE(__LINE__) ")"

#define __METAL_PP_NARG(...)  __METAL_PP_NARG_(__VA_ARGS__,__METAL_PP_RSEQ_N())
#define __METAL_PP_NARG_(...) __METAL_PP_ARG_N(__VA_ARGS__)
#define __METAL_PP_ARG_N(_1, _2, _3, _4, _5, _6, _7, _8, _9,_A, _B, _C, _D, _E, _F, N,...) N
#define __METAL_PP_RSEQ_N() F, E, D, C, B, A, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0

#define METAL_OVERLOAD(Macro, Args...) METAL_CONCAT(Macro, __METAL_PP_NARG(Args)) (Args)

#endif
