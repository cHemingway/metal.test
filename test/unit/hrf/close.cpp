
#include <array>

#include <metal/unit.hpp>

int main(int argc, char * argv[])
{

    METAL_ASSERT_CLOSE(1., .9, .1);
    METAL_EXPECT_CLOSE(1., .9, .09);

    METAL_ASSERT_CLOSE_PERCENT(100, 90, 9);
    METAL_EXPECT_CLOSE_PERCENT(100, 90, 10);

    METAL_ASSERT_CLOSE_RELATIVE(2., 1.8, 0.1);
    METAL_EXPECT_CLOSE_RELATIVE(2., 1.5, 0.25);

    std::array<int, 3> a1 = {1,2,3};
    std::array<double, 3> a2 = {1.1, 1.8, 2.7};

    METAL_ASSERT_CLOSE_RANGED(a1.begin(), a1.end(), a2.begin(), a2.end(), 0.3);
    METAL_EXPECT_CLOSE_RANGED(a1.begin(), a1.end(), a2.begin(), a2.end(), 0.2);

    METAL_ASSERT_CLOSE_RELATIVE_RANGED(a1.begin(), a1.end(), a2.begin(), a2.end(), 0.1);
    METAL_EXPECT_CLOSE_RELATIVE_RANGED(a1.begin(), a1.end(), a2.begin(), a2.end(), 0.05);

    METAL_ASSERT_CLOSE_PERCENT_RANGED(a1.begin(), a1.end(), a2.begin(), a2.end(),  5);
    METAL_EXPECT_CLOSE_PERCENT_RANGED(a1.begin(), a1.end(), a2.begin(), a2.end(), 10);


    return METAL_REPORT();
}
