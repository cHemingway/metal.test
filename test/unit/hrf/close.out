starting test execution
close.cpp(9) assertion succeeded [equality]: .9 == 1. +/- .1; [0.89999999999999991 == 1 +/- 0.099999999999999992]
close.cpp(10) expectation failed [equality]: .9 == 1. +/- .09; [0.89999999999999991 == 1 +/- 0.089999999999999997]
close.cpp(12) assertion failed [equality]: 90 == 100 +/- 9%; [ 90 == 100 +/- 9 % ]
close.cpp(13) expectation succeeded [equality]: 90 == 100 +/- 10%; [ 90 == 100 +/- 10 % ]
close.cpp(15) assertion succeeded [equality]: 1.8 == 2. +/- 0.1~; [ 1.7999999999999998 == 2 +/- 0.099999999999999992 ~ ]
close.cpp(16) expectation succeeded [equality]: 1.5 == 2. +/- 0.25~; [ 1.5 == 2 +/- 0.25 ~ ]
close.cpp(21) report: entering ranged test [{[a1.begin(), a1.end()], [a2.begin(), a2.end()]}]
close.cpp(21) assertion succeeded [equality]: **range**[0]; [1.1000000000000001 == 1 +/- 0.29999999999999999]
close.cpp(21) assertion succeeded [equality]: **range**[1]; [1.8 == 2 +/- 0.29999999999999999]
close.cpp(21) assertion succeeded [equality]: **range**[2]; [2.7000000000000002 == 3 +/- 0.29999999999999999]
close.cpp(21) exiting ranged test: { executed : 3, warnings : 0, errors : 0}
close.cpp(22) report: entering ranged test [{[a1.begin(), a1.end()], [a2.begin(), a2.end()]}]
close.cpp(22) expectation succeeded [equality]: **range**[0]; [1.1000000000000001 == 1 +/- 0.19999999999999998]
close.cpp(22) expectation succeeded [equality]: **range**[1]; [1.8 == 2 +/- 0.19999999999999998]
close.cpp(22) expectation failed [equality]: **range**[2]; [2.7000000000000002 == 3 +/- 0.19999999999999998]
close.cpp(22) exiting ranged test: { executed : 3, warnings : 1, errors : 0}
close.cpp(24) report: entering ranged test [{[a1.begin(), a1.end()], [a2.begin(), a2.end()]}]
close.cpp(24) assertion succeeded [equality]: **range**[0]~; [ 1.1000000000000001 == 1 +/- 0.099999999999999992 ~ ]
close.cpp(24) assertion succeeded [equality]: **range**[1]~; [ 1.8 == 2 +/- 0.099999999999999992 ~ ]
close.cpp(24) assertion succeeded [equality]: **range**[2]~; [ 2.7000000000000002 == 3 +/- 0.099999999999999992 ~ ]
close.cpp(24) exiting ranged test: { executed : 3, warnings : 0, errors : 0}
close.cpp(25) report: entering ranged test [{[a1.begin(), a1.end()], [a2.begin(), a2.end()]}]
close.cpp(25) expectation failed [equality]: **range**[0]~; [ 1.1000000000000001 == 1 +/- 0.049999999999999996 ~ ]
close.cpp(25) expectation failed [equality]: **range**[1]~; [ 1.8 == 2 +/- 0.049999999999999996 ~ ]
close.cpp(25) expectation failed [equality]: **range**[2]~; [ 2.7000000000000002 == 3 +/- 0.049999999999999996 ~ ]
close.cpp(25) exiting ranged test: { executed : 3, warnings : 3, errors : 0}
close.cpp(27) report: entering ranged test [{[a1.begin(), a1.end()], [a2.begin(), a2.end()]}]
close.cpp(27) assertion failed [equality]: **range**[0]%; [ 1.1000000000000001 == 1 +/- 5 % ]
close.cpp(27) assertion failed [equality]: **range**[1]%; [ 1.8 == 2 +/- 5 % ]
close.cpp(27) assertion failed [equality]: **range**[2]%; [ 2.7000000000000002 == 3 +/- 5 % ]
close.cpp(27) exiting ranged test: { executed : 3, warnings : 0, errors : 3}
close.cpp(28) report: entering ranged test [{[a1.begin(), a1.end()], [a2.begin(), a2.end()]}]
close.cpp(28) expectation succeeded [equality]: **range**[0]%; [ 1.1000000000000001 == 1 +/- 10 % ]
close.cpp(28) expectation succeeded [equality]: **range**[1]%; [ 1.8 == 2 +/- 10 % ]
close.cpp(28) expectation succeeded [equality]: **range**[2]%; [ 2.7000000000000002 == 3 +/- 10 % ]
close.cpp(28) exiting ranged test: { executed : 3, warnings : 0, errors : 0}
free tests : { executed : 12, warnings : 3, errors : 2}
full test report: { executed : 12, warnings : 3, errors : 2}